pkg/releaselock/controller.go

```diff
@@ -0,0 +1,203 @@
+		wait.Forever(func() {
+			cmList, err := c.client.CoreV1().ConfigMaps(util.ConfigMapNamespace).List(ctx, metav1.ListOptions{})
+			if err != nil {
+				klog.Errorf("Failed to list configmap in namespace %s: %v", util.ConfigMapNamespace, err)
+			}
+			klog.Infof("Listed %d configmaps in namespace %s", len(cmList.Items), util.ConfigMapNamespace)
+			for _, cm := range cmList.Items {
+				// Filter out root ca.
+				if cm.Name == rootCA {
+					continue
+				}
+				if err := c.syncLockInfo(ctx, &cm); err != nil {
+					klog.Errorf("Failed to sync lock info for configmap %s/%s: %v", cm.Namespace, cm.Name, err)
+				}
+			}
+		}, defaultSyncPeriod)
+	}
+
+	rl, err := resourcelock.New(
+		resourcelock.LeasesResourceLock,
+		c.leaderElectionNamespace,
+		leaseName,
+		nil,
+		c.client.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity: c.id,
+		})
+	if err != nil {
+		klog.Fatalf("Error creating resourcelock: %v", err)
+	}
+
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: c.leaseDuration,
+		RenewDeadline: c.renewDeadline,
+		RetryPeriod:   c.retryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
+				klog.Fatalf("%s no longer the leader, staying inactive.", c.id)
+			},
+		},
+	})
+}
+
+func (c *LockReleaseController) syncLockInfo(ctx context.Context, cm *corev1.ConfigMap) error {
+	nodeName, err := util.GKENodeNameFromConfigMap(cm)
+	if err != nil {
+		klog.Errorf("Failed to get GKE node name from configmap %s/%s: %v", cm.Namespace, cm.Name, err)
+		return err
+	}
+
+	klog.Infof("Getting GKE Node %s from API Server", nodeName)
+	node, err := c.client.CoreV1().Nodes().Get(ctx, nodeName, metav1.GetOptions{})
+	if err != nil && !apiError.IsNotFound(err) {
+		klog.Errorf("Failed to get node %s: %v", nodeName, err)
+		return err
+	}
+
+	data := cm.DeepCopy().Data
+	recentCM := cm.DeepCopy()
+	for key, filestoreIP := range data {
+		_, _, _, _, gkeNodeID, gkeNodeInternalIP, err := util.ParseConfigMapKey(key)
+		if err != nil {
+			klog.Errorf("Failed to parse configmap key %s: %v", key, err)
+			continue
+		}
+		klog.Infof("Verifying GKE node %s with nodeId %s nodeInternalIP %s exists or not", nodeName, gkeNodeID, gkeNodeInternalIP)
+		if c.verifyNodeExists(node, gkeNodeID, gkeNodeInternalIP) {
+			klog.Infof("GKE node %s with nodeId %s nodeInternalIP %s still exists in API server, skip lock info reconciliation", nodeName, gkeNodeID, gkeNodeInternalIP)
+			continue
+		}
+		klog.Infof("GKE node %s with nodeId %s nodeInternalIP %s no longer exists, releasing lock for GKE node IP %s Filestore IP %s", nodeName, gkeNodeID, gkeNodeInternalIP, gkeNodeInternalIP, filestoreIP)
+		if err := ReleaseLock(filestoreIP, gkeNodeInternalIP); err != nil {
+			klog.Errorf("Failed to release lock: %v", err)
+			continue
+		}
+		recentCM, err = util.RemoveKeyFromConfigMap(ctx, recentCM, key, c.client)
```

Comment:
note for testing: let's say we remove lock successfully, but fail to remove configMap key. To ensure the retry behavior works, we should test the scenario manually to check a no-op lock release call is handled gracefully by ReleaseLock()

Reply:
I agree Get+Update() should reduce the probability of version conflicts. Given O(confimaps) is expected to be small. But we cannot guarantee that it will always succeed. And this means, lock release needs to be idempotent anyway. correct?

Reply:
resolving offline, we'll go with `get+update` approach